bin/persona-gmail-bridge.js

```diff
@@ -280,43 +263,65 @@ app.get('/authenticate/forward', validate({ email: 'gmail' }),
 
 app.get('/authenticate/verify', function (req, res) {
   // Check input precondition:
+  // The bridge should return the expected state token.
+  if (! (req.query.state &&
+         req.session.state &&
+         req.query.state === req.session.state)) {
+    logger.error('Invalid or missing state');
+    statsd.increment('authentication.oauth.failure.no_state');
+    return res.status(400).render('error',
+      { title: req.gettext('Error'), errorInfo: 'Invalid or missing state.' });
+  }
+  // a state token is only good for one request
+  delete req.session.state;
+
+  // Fail early if the user hit cancel in the OAuth dialog
+  if (req.query.error === 'access_denied') {
+      logger.info('User cancelled during oauth dialog');
+      statsd.increment('authentication.oauth.failure.cancelled');
+      return res.render('authenticate_finish',
+        { title: req.gettext('Loading...'), success: false });
+  }
+
+  var code = req.query.code;
+  if (!code) {
+    logger.error('Invalid or missing code');
+    statsd.increment('authentication.oauth.failure.no_code');
+    return res.status(400).render('error',
+      { title: req.gettext('Error'), errorInfo: 'Invalid or missing code.' });
+  }
+
+  // Check input precondition:
   // Session should include a valid email address that the user is claiming.
   if (!email.valid(req.session.claimed)) {
     logger.error('Invalid or missing claimed email');
-    statsd.increment('authentication.openid.failure.no_claim');
+    statsd.increment('authentication.oauth.failure.no_claim');
     return res.status(400).render('error',
       { title: req.gettext('Error'), errorInfo: 'Invalid or missing claim.' });
   }
 
-  // Check input precondition:
-  // OpenID params must include only one email address, and it must be signed.
-  if (!oidTool.validParams(req.query)) {
-    return res.status(401).render('error',
-      { title: req.gettext('Error'), errorInfo: 'Email not signed.' });
-  }
-
-  openidRP.verifyAssertion(req, function (error, result) {
+  google.tradeCodeForEmail(code, function (error, result) {
     if (error && error.message === 'Authentication cancelled') {
```

Comment:
We should safely be able to remove this branch. We were previously passing in the request to `verifyAssertion`, which would check for errors and surface them.